### PR TITLE
[5.4] Oracle Error Reconnect Handling

### LINKED ISSUE #364: [5.4/5] Oracle error codes for reconnect handling
Reconnect handling currently uses laravel defaults.

See https://github.com/illuminate/database/blob/master/DetectsLostConnections.php#L20

But oracle has different codes/error messages and we have seen this problem in production.

This needs to be added and changable via config. 
A Co-Worker of mine is currently fixing this and will make a PR.

### PR DESCRIPTION
**Pull request by Stoffo**
This PR is fixing the Reconnect Methods to react to Oracles specific Error Codes

Fixes #364

### CODE REVIEW

File: src/Oci8/Oci8Connection.php [hunk 1]
```diff
@@ -442,4 +446,28 @@ public function addBindingsToStatement(PDOStatement $stmt, array $bindings)
...
+            'ORA-03135',    //Connection lost contact
+            'ORA-12170',    //Connect timeout occurred
+            'ORA-12537',    //Connection closed
+            'ORA-27146',    //Post/wait initialization failed
+            'ORA-25408',    //Can not safely replay call
+            'ORA-56600',    //Illegal Call
+        ];
+
+        if (is_array($this->config['options'][static::RECONNECT_ERRORS])) {
+            $lostConnectionErrors = array_merge($this->config['options'][static::RECONNECT_ERRORS]);
```
Comment (mstaack) [inline src/Oci8/Oci8Connection.php]: i guess there is something missing in this array_merge call

**Comment by mstaack** [comment]
@yajra ping

File: src/Oci8/Oci8Connection.php [hunk 2]
```diff
@@ -442,4 +446,35 @@ public function addBindingsToStatement(PDOStatement $stmt, array $bindings)
 
         return $stmt;
     }
+
+    public function causedByLostConnection(Exception $e)
```
Comment (yajra) [inline src/Oci8/Oci8Connection.php]: Can you please add the doc blocks.

File: src/Oci8/Oci8Connection.php [hunk 3]
```diff
@@ -2,23 +2,27 @@
...
+use PDO;
+use PDOStatement;
 use Yajra\Oci8\Query\Grammars\OracleGrammar as QueryGrammar;
+use Yajra\Oci8\Query\OracleBuilder as QueryBuilder;
 use Yajra\Oci8\Query\Processors\OracleProcessor as Processor;
 use Yajra\Oci8\Schema\Grammars\OracleGrammar as SchemaGrammar;
+use Yajra\Oci8\Schema\OracleBuilder as SchemaBuilder;
+use Yajra\Oci8\Schema\Sequence;
+use Yajra\Oci8\Schema\Trigger;
+use Yajra\Pdo\Oci8\Statement;
```
Comment (yajra) [inline src/Oci8/Oci8Connection.php]: I think the cs for laravel import is it was arranged from shortest to longest name. Might be the reason why styleci is failing.

**Comment by yajra** [comment]
The PR looks good. Need to test though on actual project and need some minor changes. Thanks!

**Comment by Stoffo** [comment]
@yajra Fixed all the issues, you can take a look again :+1:

**Comment by yajra** [comment]
Released on v5.4.19. Can you please also send another PR for 5.5 or you prefer that I just cherry pick/merge this commit to 5.5? Thanks!

**Pull request merged**